### DISCUSSION TITLE
OD-beta and EGL

### DISCUSSION THREAD

**Discussion by a-dekker** [eduardofilo/RG350_adam_image#255] [Ports]
Some games use EGL (https://en.wikipedia.org/wiki/EGL_(API)) to get the GLESv2 render stuff to the screen. https://sourceforge.net/projects/eglport/ is also often used in existing Pandora/GCW0 ports (a set of functions to interface EGL to SDL-raw framebuffers). But Eglport usage results in: Fails to initialize EGL display "Unable to query for available configs, found 0". Quake 3 is an example of that; it did work in the old OpenDingux environment.

Does anybody know what the EGL state is in OD-beta? All I get is "bad surface/Can't initialize EGL context/OpenGLES: eglSwapBuffers failed!" like errors during execution.

**Comment by eduardofilo** [comment]
I asked @plrguez, who is the person I know most knowledgeable about the topic, and he told me that he couldn't use EGL either and had to switch to SDL with GLES2. He also referred me to this issue where he talks about it.
https://github.com/XProger/OpenLara/issues/365#issuecomment-962464811

**Comment by a-dekker** [comment]
A bit sad to hear, but thanks for contacting Pirguez for providing a bit more inside on the state of EGL.

**Reply by plrguez** [reply]
Hi, I've taken time this weekend to review this and have made some progress.

kmscube is working great on OD-beta (it is installed by default and can be run from an ssh session). So I had take a look at the code https://gitlab.freedesktop.org/mesa/kmscube/

So to test it I have taken some code from kmscube and successfully compiled OpenLara using EGL. https://github.com/plrguez/OpenLara/commit/7b4a6ba560649592ecd15f01e2fe3515092cf9d4
I have used the drm-legacy implementation as base. The code could be simplified more. 

The opk inside this zip file ([OpenLara.zip](https://github.com/eduardofilo/RG350_adam_image/files/8886861/OpenLara.zip)) is not a release, some features are missing, and only load the game files from `$HOME/.openlara`. Performance is similar to SDL2/GLES2 release (tested on RG350M).

So EGL is working ok in OD-beta and with some adaptations other EGL ports could be compiled for it.

**Reply by a-dekker** [reply]
Thanks for investigating. Not sure if I will be able to map all the relevant changes to other ports, but I will definitely have a further look at it.

**Comment by a-dekker** [comment]
@plrguez if I could just do one request for a fixed EGL part in a game it would be the bullet-hell shooter game rRootage (https://github.com/jamesofarrell/rRootage-for-Handhelds). It is a bit like Noiz2sa, but on steroids ;-) . More info here https://boards.dingoonity.org/gcw-releases/rrootage-for-gcw-zero/

I did have a look at some of your drm/kms changes in Hurricane and OpenLara, but it's not that straightforward to implement in other games (at least for me, having little knowledge regarding this low level part).

**Reply by plrguez** [reply]
Hi @a-dekker , I have included the DRM/KMS changes in rRootage code. Try the opk from [here](https://github.com/plrguez/rRootage-for-Handhelds/releases/tag/2022-10-24-od-v0.0.1)

The code for include the DRM/KMS is massive, but is always the same with little variations.

**Reply by a-dekker** [reply]
That was almost faster than the game itself ;-)
Thanks, it seems to work fine here!
